File: others/quick_sort.cpp
#include <iostream>

int arr[] = {6, 3, 1, 2, 5, 4, 9, 8, 7};

void quick_sort(int arr[], int start, int end)
{
    if (start >= end)
        return;

    int pivot = start;
    int left = pivot + 1;
    int right = end;

    // 피벗 기준 분할
    while (left <= right)
    {
        while (left <= end && arr[left] <= arr[pivot])
        {
            left++;
        }

        while (right > start && arr[right] >= arr[pivot])
        {
            right--;
        }

        if (left > right)
        {
            int tmp = arr[pivot];
            arr[pivot] = arr[right];
            arr[right] = tmp;
        }
        else
        {
            int tmp = arr[left];
            arr[left] = arr[right];
            arr[right] = tmp;
        }
    }
    quick_sort(arr, start, right-1);
    quick_sort(arr, right+1, end);

}

int main(int argc, char const *argv[])
{
    int len = sizeof(arr)/sizeof(int);
    quick_sort(arr, 0, len - 1);

    for (int num : arr){
        std::cout << num << " ";
    }
    return 0;
}

File: others/quick_sort.js
const arr = [5, 3, 1, 4, 2, 6, 8, 7, 9];

const quickSort = (arr, start, end) => {
  if (start >= end) {
    return;
  }

  let pivot = start;
  let left = start + 1;
  let right = end;

  while (left <= right) {
    while (left <= end && arr[left] <= arr[pivot]) {
      left++;
    }

    while (right > start && arr[right] >= arr[pivot]) {
      right--;
    }

    if (left > right) {
      const tmp = arr[pivot];
      arr[pivot] = arr[right];
      arr[right] = tmp;
    } else {
      const tmp = arr[left];
      arr[left] = arr[right];
      arr[right] = tmp;
    }
  }

  quickSort(arr, start, right - 1);
  quickSort(arr, right + 1, end);
};

quickSort(arr, 0, arr.length - 1);
console.log(arr);
